src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java

```diff
@@ -70,31 +75,138 @@ protected ActionResult execute() throws EntityDoesNotExistException{
 //      unless the the page is reloaded with "?testdata=false"  or simply reloaded with this parameter omitted.       
         boolean ifShowTestData = getRequestParamAsBoolean("testdata");
         
-        
-        if (pageChange != null && pageChange.equals("true")) {
-            //Reset the offset because we are performing a new search, so we start from the beginning of the logs
-            offset = null;
-        }
         if (filterQuery == null) {
             filterQuery = "";
         }
         //This is used to parse the filterQuery. If the query is not parsed, the filter function would ignore the query
         data.generateQueryParameters(filterQuery);
         
-        LogQuery query = buildQuery(offset, data);
-        
-        List<ActivityLogEntry> logs = getAppLogs(query, data);
-        
-        data.init(offset, filterQuery, ifShowAll, ifShowTestData, logs);
+        List<ActivityLogEntry> logs = null;
+        if (data.isFromDateSpecifiedInQuery()) {
+            logs = searchLogsWithExactTimePeriod(data);
+        } else {
+            if (!searchTimeOffset.isEmpty()) {
+                data.setToDate(Long.parseLong(searchTimeOffset));
+            }
+            logs = searchLogsWithTimeIncrement(data);
+        }
+        generateStatusMessage(data, logs);
+        data.init(ifShowAll, ifShowTestData, logs);
         
-        if (offset == null) {
+        if (searchTimeOffset.isEmpty()) {
             return createShowPageResult(Const.ViewURIs.ADMIN_ACTIVITY_LOG, data);
         }
         
         return createAjaxResult(data);
     }
     
-    private LogQuery buildQuery(String offset, AdminActivityLogPageData data) {
+    private void generateStatusMessage(AdminActivityLogPageData data, List<ActivityLogEntry> logs) {
+        String status = "Total Logs gone through in last search: " + totalLogsSearched + "<br>";
+        status += "Total Relevant Logs found in last search: " + logs.size() + "<br>";
+        
+        long earliestSearchTime = data.getFromDate();
+        ActivityLogEntry earliestLogChecked = null;
+        if (!logs.isEmpty()) {
+            earliestLogChecked = logs.get(logs.size() - 1);
+        }
+        //  if the search space is limited to a certain log
+        if ((logs.size() >= RELEVANT_LOGS_PER_PAGE) && (earliestLogChecked != null)) {
+            earliestSearchTime = earliestLogChecked.getTime();
+        }
+        
+        double targetTimeZone = Const.DOUBLE_UNINITIALIZED;
+        if (data.isPersonSpecified()) {
+            String targetUserGoogleId = data.getPersonSpecified();
+            targetTimeZone = getLocalTimeZoneForRequest(targetUserGoogleId, "");
+        }
+        
+        if (targetTimeZone == Const.DOUBLE_UNINITIALIZED) { // if no person is specified or the person doesn't really exist
+            if ((logs.size() >= RELEVANT_LOGS_PER_PAGE) && (earliestLogChecked != null)) {
```

Comment:
Here the `earliestSearchTime` can be either `data.getFromDate()` or `earliestLogChecked.getTime()`. If the `earliestSearchTime` is `earliestLogChecked.getTime()`, it should use the timezone found from the log, otherwise the admin timezone should be used instead. That's why the condition in line 113 and line 124 must be identical.